### ISSUE TITLE
wazuh-states-vulnerabilities keeps data for deleted agent

### ISSUE THREAD

**Issue by ndonkersloot** [wazuh/wazuh#24942]
|Wazuh version|Component|Install type|Install method|Platform|
|4.8.0|vulnerability-detector|Manager|docker|Fedora CoreOS|

I'm using the new vulnerability dashboard (which is awesome, thanks a lot!) and noticed critical vulnerability reports for agents that I have deleted. If I discover inside the wazuh-states-vulnerabilities-* index pattern I still see events of this agent but it is correctly removed in the rest of Wazuh.

Is this behaviour expected?
I think it would be logical to also delete all history in the wazuh-states-vulnerabilities-* on delete of a agent.

What is the correct way to manually delete events of a specific agent inside this index?

**Comment by sebasfalcone** [comment]
@MiguelazoDS Please perform an RCA on this issue

**Comment by MiguelazoDS** [comment]
Hello @ndonkersloot, 

I did a quick test and I was not able to reproduce it. 

What I did was: 

- Deployed single-node manager
- Added a centos9 agent
- Wait until vulnerabilities are scanned and displayed in the dashboard
- Stop the agent (so it does not register again) 
- Remove the agent using the manage-agent tool. 

Was that the same procedure you followed?

**Comment by ndonkersloot** [comment]
> Hello @ndonkersloot,
> 
> I did a quick test and I was not able to reproduce it.
> 
> What I did was:
> 
>     * Deployed single-node manager
> 
>     * Added a centos9 agent
> 
>     * Wait until vulnerabilities are scanned and displayed in the dashboard
> 
>     * Stop the agent (so it does not register again)
> 
>     * Remove the agent using the manage-agent tool.
> 
> 
> Was that the same procedure you followed?

Excuses me, i should've added this in my original post.
All steps, except the last are the same. I've deleted the agent via the API using the API console from the webinterface.

I can't find the history for the exact API call I did, but I'm sure it is something like this:

DELETE /agents?status=disconnected&agents_list=001

**Comment by MiguelazoDS** [comment]
Thanks for the info! the outcome is the same regardless of the two methods. I think this is working as designed but I want to discuss it with my team because it is confusing. Visually we can only see that the indexer is not updating as expected and there is remaining data in the dashboard. 

After the scan finishes it does not update the index
![2024-07-26_16-31](https://github.com/user-attachments/assets/3bf171c2-e9fd-4244-86d9-7719f5307780)

Because of this control
```console
2024/07/26 16:08:24 indexer-connector[100972] indexerConnector.cpp:129 at abuseControl(): DEBUG: Agent '001' sync omitted due to abuse control.
```

What I see is that after a default value of 30 minutes (which is also too big), I don't see the vulnerabilities in the index. 

Only after restarting the manager (that causes the thread to end its tasks) do I see the vulnerabilities indexed 

![image](https://github.com/user-attachments/assets/59883074-f672-4544-a469-8819b2e840ff)

Did you restart your manager? do you still have the issue? 

Once I have more info about this I'll update with a new comment. Thanks!

**Comment by ndonkersloot** [comment]
@MiguelazoDS 

Thank you for your detailed response. I currently don't have the time to log-in to give more specific details but I'm certain the manager is restarted after the deletion of the agents. At least once via the web interface because of config changes to the manager, and at least once the full stack by stopping all containers specified in the docker-compose file.

I'm sure I have more disconnected agents that I can delete to see if I can reproduce this consistently. I currently have this for 3 agents that I know of which has still lingering vulnerabilities after deletion. These agents got deleted at least one and a half week ago. 

Probably redundant but to summerize my problem:
I deleted some agents using the API. In the agent lists these agents are deleted. In the wazuh-alerts-* index pattern all documents of these agents are deleted. In the new vulnerability dashboard vulnerabilities of these agents still show up, as well as in the wazuh-states-vulnerabilities-* index pattern. I've restarted the manager and even the full stack without a difference.

**Comment by MiguelazoDS** [comment]
It's ok, the information you provided was really useful, I found some things that may need some discussion and adjustment. Probably the environment ended in an unstable state for some reason I need to figure out. 

Answering your question to overcome this behavior. 

the indexed documents are like this 

```json
      {
        "_index": "wazuh-states-vulnerabilities-jammy",
        "_id": "001_02eb48849999f4d3ea1cca745fe7f05cf5bfdee9_CVE-2022-27943",
        "_score": 1,
        "_source": {
          "agent": {
            "id": "001",
            "name": "centos9",
            "type": "wazuh",
            "version": "v4.8.1"
          },
          "host": {
            "os": {
              "full": "CentOS Stream 9",
              "kernel": "5.14.0-378.el9.x86_64",
              "name": "CentOS Stream",
              "platform": "centos",
              "type": "centos",
              "version": "9"
            }
 ```
            
you can delete them with this query 

```console
root@jammy:/home/vagrant/wazuh# curl -k -u admin:admin -XDELETE https://127.0.0.1:9200/wazuh-states-vulnerabilities-jammy/_doc/001_02eb48849999f4d3ea1cca745fe7f05cf5bfdee9_CVE-2022-27943 | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   228  100   228    0     0  17880      0 --:--:-- --:--:-- --:--:-- 19000
{
  "_index": "wazuh-states-vulnerabilities-jammy",
  "_id": "001_02eb48849999f4d3ea1cca745fe7f05cf5bfdee9_CVE-2022-27943",
  "_version": 2,
  "result": "deleted",
  "_shards": {
    "total": 1,
    "successful": 1,
    "failed": 0
  },
  "_seq_no": 8937,
  "_primary_term": 1
}
```        

Since this is not practical and it seems it does not accept wildcards, another option is _delete_by_query 

https://opensearch.org/docs/latest/api-reference/document-apis/delete-by-query/

But I couldn't find a way to do that. 

I would use something like this to get all the ids that match with "001_" (my case) 

`curl -k -u admin:admin https://127.0.0.1:9200/wazuh-states-vulnerabilities-*/_search | jq '.hits.hits[]|select(._id|test("001_"))|._id'
`

and repeat the previous command with every single id found in the previous query. 

`curl -k -u admin:admin -XDELETE https://127.0.0.1:9200/wazuh-states-vulnerabilities-jammy/_doc/${ID_FOUND_PREVIOUS_COMMAND}`

**Comment by ndonkersloot** [comment]
Thank you!
On Monday I'll delete another agent to check if this behavior reoccurs. For now i'll delete documents manually like you described.

**Comment by ndonkersloot** [comment]
I'm sorry I didn't respond on monday to test if I can reproduce consistently. I've been sick this entire week so far.

Of course I will report once I feel well again.

**Comment by sebasfalcone** [comment]
Hope you recover soon! We will await for your report

**Comment by MiguelazoDS** [comment]
# Update 

> [!NOTE]
> Unable to reproduce any error 

I tested again on 4.8.0 and 4.8.1 but could not reproduce any issue during the deletion of the agent or the behaviors mentioned above. 

Starting with an empty indexer and updated database
![image](https://github.com/user-attachments/assets/729bb432-6614-43cd-914a-5850a7c15c49)

It keeps indexing 
![image](https://github.com/user-attachments/assets/d2f68147-3483-4cab-8c44-93c8893811cd)
![image](https://github.com/user-attachments/assets/90beefac-420f-4cf9-9c7f-d1f187a4c1cd)

I delete the agent and the indexed values start decreasing until every indexed vulnerability is removed.
![image](https://github.com/user-attachments/assets/92638f23-6ca8-45ab-9855-199382856a8f)
![image](https://github.com/user-attachments/assets/0082c778-4614-47bc-9c4f-d72066c86735)

**Comment by sebasfalcone** [comment]
## Update
One of the users who reported this issue confirmed that the cause was a lack of disk space on the indexer side

## Conclusion
We will close this issue as we were unable to replicate the issue

**Comment by ndonkersloot** [comment]
@MiguelazoDS 
I had severe tonsilitis and was out for some time. Sorry for my (very) delayed response.
We currently run 4.8.2 and haven't reproduced this but stay wary of it and if I can reproduce it consistently I'll create a new issue.
Thank you very much for looking into this.

I want to share here how I cleaned the lingering docs, maybe it'll help someone.
First I opened the vulnerability details so I could see the agent id, let's say it was '001'.
Then I double checked using the Dev Tools in the server management sub menu if the agent is really gone.

`GET /agents?agents_list=001`

Then under the index management dev tools I retrieve the docs to double check if these really are the ones I want to delete:

```
GET wazuh-states-vulnerabilities-wazuh/_search
{
  "query": {
    "match": {
      "agent.id": "001"
    }
  }
}

```

And delete all docs at once like so:
```

POST wazuh-states-vulnerabilities-wazuh/_delete_by_query
{
  "query": {
    "match": {
      "agent.id": "001"
    }
  }
}

```
